lib/zz.mli

```diff
@@ -0,0 +1,183 @@
+(** {1 ZLIB layer.}
+
+    ZLIB is a standard on top of RFC1951. It uses [dd] implementation with LZ77
+   compression algorithm. Module provides non-blocking streaming codec to
+   {{:#decode}decode} and {{:#encode}encode} ZLIB encoding. It can efficiently
+   work payload by payload without blocking IO. *)
+
+module Bigarray = Bigarray_compat
+(** MirageOS compatibility. *)
+
+type bigstring = (char, Bigarray.int8_unsigned_elt, Bigarray.c_layout) Bigarray.Array1.t
+(** The type for [bigstring]. *)
+
+type window = Dd.window
+(** The type for windows. *)
+
+val io_buffer_size : int
+(** IO_BUFFER_SIZE 4.0.0 *)
+
+(** {2:decode ZLIB Decoder.}
+
+    Despite [dd], [zz] provides a referentially transparent {!M.decoder}. The client
+   must use [decoder] given {b by} {!M.decode} instead [decoder] given {b to}
+   {!M.decode}. A common use of [zz] is:
+
+     {[
+let rec go d0 = match M.decod d0 with
+  | `Await d1 -> ... go d1
+  | `Flush d1 -> ... go d1
+  | _ -> ... in
+     ]} *)
+
+module M : sig
+  type decoder
+  (** The type for decoders. *)
+
+  type src = [ `Channel of in_channel | `String of string | `Manual ]
+  (** The type for input sources. With a [`Manual] source the client must
+     provide input with {!src}. With [`String] or [`Channel] source the client
+     can safely discard [`Await] case (with [assert false]). *)
+
+  type decode = [ `Await of decoder | `Flush of decoder | `End of decoder | `Malformed of string ]
+
+  val decoder : src -> o:bigstring -> allocate:(int -> window) -> decoder
+  (** [decoder src ~o ~allocate] is a decoder that inputs from [src].
+
+      {b Output buffer.}
+
+      [zz], as [dd], uses [o] buffer as internal buffer to store output. We
+     recommend to allocate an {!io_buffer_size} buffer as output buffer. Then,
```

Comment:
I did not have benchmark about that but you should be right about this detail where a `bigstring` is located to the C heap (and allocated by `malloc`). Optimization by _page-size_ can, in fact, improve I/O. I will try and return to you results.

Reply:
Arf, in `lzbench` we can not allocate by ourselves buffers and choose how many bytes we want to allocate. I will see if we have a difference only between `pipe with page` / `pipe without page`

Reply:
So I compare time needed by `pipe.exe` on a huge compressed file (286M) and I did see any differences (~36s to decompress all) with `65536 - 3*8`.

Reply:
Did or did not?
I suspect it will be a bit hard to benchmark; it would depend heavily on memory pressure and TLB utilization, since that is when the extra page will be expensive.

Reply:
**did not**

Reply:
But, even if we advise the client to allocate `io_buffer_size` bytes, this parameter still is under the control of the user.